sys/iec62056/iec62056_21.c

```diff
@@ -0,0 +1,391 @@
+            return false;
+        }
+        begin++;
+    }
+    return true;
+}
+
+bool iec62056_21_dataset_is_register(const iec62056_21_dataset_t *dataset)
+{
+    const char *end_of_register = &dataset->value[dataset->value_length];
+    const char *fractional_part = memchr(dataset->value, '.', dataset->value_length);
+    const char *register_unit = memchr(dataset->value, '*', dataset->value_length);
+
+    if (dataset->value_length == 0) {
+        return false;
+    }
+
+    if (!register_unit) {
+        register_unit = end_of_register;
+    }
+
+    if (!fractional_part) {
+        fractional_part = register_unit;
+    }
+
+    if (!_register_part_is_numeric(dataset->value, fractional_part)) {
+        return false;
+    }
+
+    if (fractional_part == end_of_register) {
+        return true;
+    }
+    fractional_part++;
+
+    return _register_part_is_numeric(fractional_part, register_unit);
+}
+
+int iec62056_21_dataset_parse_register_to_decfrac(const iec62056_21_dataset_t *dataset,
+                                                  int64_t *mantissa, int64_t *scaling)
+{
+    /* parse a string with format '001234.567*kWh' to components */
+    if (dataset->value_length == 0) {
+        return -1;
+    }
+
+    const char *reg = dataset->value;
+
+    const char *end_of_register = &reg[dataset->value_length];
+    const char *frac = memchr(reg, '.', dataset->value_length);
+    const char *unit = memchr(
+        frac ? frac : reg,
+        '*',
+        frac ? (size_t)(end_of_register - frac) : dataset->value_length);
+
+    if (!unit) {
+        unit = end_of_register;
+    }
+
+    bool negative = false;
+
+    if ((dataset->value_length >= 2) && (*reg == '-')) {
+        negative = true;
+        reg++;
+    }
+
+    size_t num_len = frac ? frac - reg : unit - reg;
+
+    for (size_t i = 0; i < num_len; i++) {
+        if (!fmt_is_digit(reg[i])) {
+            return -1;
+        }
+    }
+    uint32_t num = scn_u32_dec(reg, num_len);
+
+    *mantissa = num;
+    *scaling = 0;
+
+    if (frac) {
+        frac++;
+        size_t fraction_len = unit - frac;
+        unsigned power = 1;
+        for (unsigned i = 0; i < fraction_len; i++) {
+            if (!fmt_is_digit(frac[i])) {
+                return -1;
+            }
+            power *= 10;
+        }
+        uint32_t fraction = scn_u32_dec(frac, fraction_len);
+        *mantissa *= power;
+        *mantissa += fraction;
+        *scaling = ((int64_t)fraction_len) * -1;
+    }
+    if (negative) {
+        *mantissa *= -1;
+    }
+    return 0;
+}
+
+int iec62056_21_dataset_parse_register_to_double(const iec62056_21_dataset_t *dataset,
+                                                 double *value)
+{
+    int64_t mantissa = 0;
+    int64_t scaling = 0;
+    int res = iec62056_21_dataset_parse_register_to_decfrac(dataset, &mantissa, &scaling);
+
+    if (res < 0) {
+        return res;
+    }
+
+    *value = (double)mantissa * pow(10, scaling);
```

Comment:
```suggestion
    *value = (float)mantissa * powf(10, scaling);
```
is single precision not enough? 
A `iec62056_21_dataset_parse_register_to_float()` might be handy if we don't want to pull in soft-float on e.g. Cortex-M4F.

Reply:
Added a `iec62056_21_dataset_parse_register_to_float()`.